rust/src/adapter/datadog.rs

```diff
@@ -0,0 +1,160 @@
+            spans: vec![],
+            config,
+        };
+
+        Arc::new(Mutex::new(adapter))
+    }
+
+    pub fn new_collector(&mut self) -> usize {
+        next_id()
+    }
+
+    fn _handle_event(&mut self, event: Event, parent_id: Option<u64>) -> Option<Vec<Span>> {
+        match event {
+            Event::Func(_id, f) => {
+                let function_name = &f.name.clone().unwrap_or("unknown-name".to_string());
+                let name = format!("function-call-{}", &function_name);
+
+                let config = self.config.clone();
+                let span =
+                    Span::new(config, self.trace_id.clone(), parent_id, name, f.start, f.end);
+
+                let span_id = Some(span.span_id.clone());
+                let mut spans = vec![span];
+
+                for e in f.within.iter() {
+                    if let Some(mut s) = self._handle_event(e.to_owned(), span_id.clone()) {
+                        spans.append(&mut s);
+                    };
+                }
+
+                Some(spans)
+            }
+            Event::Alloc(_id, a) => {
+                let config = self.config.clone();
+                let span = Span::new(
+                    config,
+                    self.trace_id,
+                    parent_id,
+                    "allocation".to_string(),
+                    a.ts,
+                    a.ts,
+                );
+                Some(vec![span])
+            }
+            Event::Metadata(_id, Metadata { key, value }) => {
+                if key == "trace_id" {
+                    self.trace_id = value;
+                }
+                None
+            }
+            Event::Shutdown(_id) => {
+                // when we receive the shutdown
+                // then dump the trace to the agent
+                self.shutdown();
+                self.spans.clear();
+                None
+            },
+        }
+    }
+
+    pub async fn set_trace_id(collector: &Collector, trace_id: u64) {
+        collector
+            .set_metadata("trace_id".to_string(), trace_id)
+            .await;
+    }
+}
+
+impl Adapter for DatadogAdapter {
+    // flush any remaning spans
+    fn shutdown(&self) {
+        let mut dtf = DatadogFormatter::new();
+        let mut trace = Trace::new();
+        for span in &self.spans {
+            trace.push(span.clone());
+        }
+        dtf.traces.push(trace);
+
+
+        let host = Url::parse(&self.config.agent_host).unwrap();
+        let url = host.join("/v0.3/traces").unwrap().to_string();
+        let j = json!(&dtf.traces);
+        let body = serde_json::to_string(&j).unwrap();
+
+        let response = ureq::post(&url)
+            .set("Content-Type", "application/json")
+            .send_string(&body);
+
+        // Check the response status
+        if response.is_ok() {
+            println!("Request was successful!");
+        } else {
+            println!("Request failed with status: {:#?}", response);
```

Comment:
TODO better error handling and maybe use an async http client.

Reply:
>  maybe use an async http client.

This falls into the same category of "use what the user is using", and will require us to add features to support various async runtime libraries or if no-async, then some managed thread thing to avoid blocking.